testdata/src/go.uber.org/receivers/inference/receivers-with-inference.go

```diff
@@ -53,16 +57,21 @@ func testRecv() {
 
 // -----------------------------------
 // the below test checks for in-scope analysis of receivers. If a receiver-based call is made to an external method,
-// such as `err.Error()`, then it is treated as a normal field access of `err`, reporting an error if `err == nil`.
+// such as `err.Error()`, then it is treated with optimistic default, assuming the external method to be handling
+// nil receivers. This can potentially result in false negatives, as shown below in the example of `err.Error()`.
+// However, this is a trade-off made to avoid false positives.
 
 func (a *A) retErr() error {
 	return nil
 }
 
 func testInScope() {
+	var file *os.File
+	_, _ = file.Stat() // true negative, since `Stat()` is nil-safe
+
 	var a *A
 	err := a.retErr()
-	print(err.Error()) //want "returned as error result 0 of `retErr.*`"
+	print(err.Error()) // false negative, since `Error()` is nil-unsafe
```

Comment:
Thank you for your insightful feedback. I acknowledge your concern regarding false negatives and would like to address specific points related to extending NilAway's design for nil receivers and the potential utilization of the trusted function framework.

To provide context, NilAway is designed with optimistic defaults for external code, assuming non-nil returns and nil-safe parameters. For example, `atomic.AddUint64(nil, 1)` would lead to a nil panic upon execution due to the `nil` parameter, however, goes unreported (false negative) by NilAway due to being outside its scope. Coming to the nil receivers change in this PR, I'd like to emphasize that this adjustment aligns with this existing paradigm of NilAway. Receivers can be viewed as specialized parameters, and this PR simply extends the established design bounds of NilAway within this paradigm. I believe treating receivers in a manner consistent with parameters is both logical and in line with NilAway's core principles.

> Can we have a functionality regression check internally to see if the errors being removed by this PR are all valid false positives?

I agree and have already conducted spot checks internally, all of which confirmed the instances as false positives. However, I can perform a more thorough and systematic check to ensure the accuracy of the errors removed by this PR.

> Would it be possible to extend our trusted function framework to handle this?

While technically possible, it would involve modeling each external function individually, which could be a potentially tedious process.

> extend the trusted function framework to be a plugin system

I concur; this aligns with our internal discussions on enhancing NilAway's capabilities in the long-term.

> we probably also want to model thrift (and other IDL languages like protobuf) better

I do share your perspective on the importance of improving NilAway's handling of IDL languages like Thrift and Protobuf, leveraging their built-in nilability annotations for more precise code analysis. This would be a great long-term task to improve NilAway.

I hope this clarifies my stance on these matters. Please feel free to further discuss or provide additional insights. Happy to discuss this offline too! :)

Reply:
Thanks for the explanations!

Hmmmm, I agree that this actually aligns with the optimistic defaults in NilAway in other places, so let's have that. 😃 

The example here was from stdlib and I had a thought of actually analyzing stdlib and caching the results, so I confused myself here. 

For context, since [Go 1.20](https://go.dev/blog/go1.20#:~:text=The%20go%20tool%20no%20longer,build%20cache%2C%20like%20other%20packages.) the stdlib is no longer shipped as precompiled packages but built on demand and cached in the build cache like any other packages (meaning NilAway has a chance to "see" the stdlib code), I had a thought that since stdlib code does not change it might be good to actually analyze it (and then cache it).

But that's something for another discussion later, I think we should go for optimistic defaults as a fallback mechanism, and we can have some handlings as special cases (stdlib, thrift-generated code etc.).

Stamping the changes!

Reply:
Thanks for the context. I understand and agree with your position :) We can revisit and revamp the handling of external code in the future, but for now resort to optimistic defaults and make the handling of receivers consistent with parameters.